lib/connection/transaction.ex

```diff
@@ -0,0 +1,113 @@
+defmodule Janus.Connection.Transaction do
+  @moduledoc false
+
+  require Logger
+  @spec init_transaction_call_table() :: :ets.tab()
+  def init_transaction_call_table() do
+    :ets.new(:pending_calls, [:duplicate_bag, :private])
+  end
+
+  # Generates a transaction ID for the payload and ensures that it is unused
+  @spec generate_transaction!(:ets.tab()) :: binary
+  def generate_transaction!(pending_calls_table) do
+    transaction = :crypto.strong_rand_bytes(32) |> Base.encode64()
+
+    case :ets.lookup(pending_calls_table, transaction) do
+      [] ->
+        transaction
+
+      _ ->
+        generate_transaction!(pending_calls_table)
```

Comment:
Technically it is possible to create an infinite loop here (let's say you'll mock `:crypto` in tests). Just to be sure I'd add a limit on the number of calls.

Reply:
I added 2 layer tests for this scenario (can not insert race condition and can not generate). Mocked using explicit contracts because mockery requires all env dependency which is discouraged for library applications.